packages/govuk-frontend/src/govuk/components/password-input/password-input.puppeteer.test.js

```diff
@@ -0,0 +1,363 @@
+          await render(page, 'password-input', examples.default)
+        })
+
+        it('renders the status element', async () => {
+          const statusElement = await page.evaluate(
+            (statusClass) => document.querySelector(statusClass),
+            statusClass
+          )
+
+          expect(statusElement).not.toBeUndefined()
+        })
+
+        it('renders the status element without aria-live', async () => {
+          const statusAriaLiveAttribute = await page.evaluate(
+            (statusClass) =>
+              document.querySelector(statusClass).hasAttribute('aria-live'),
+            statusClass
+          )
+
+          expect(statusAriaLiveAttribute).toBeFalsy()
+        })
+
+        it('renders the status element empty', async () => {
+          const statusText = await page.evaluate(
+            (statusClass) =>
+              document.querySelector(statusClass).innerHTML.trim(),
+            statusClass
+          )
+
+          expect(statusText).toEqual('')
+        })
+
+        it('shows the toggle button', async () => {
+          const buttonHiddenAttribute = await page.evaluate(
+            (buttonClass) =>
+              document.querySelector(buttonClass).hasAttribute('hidden'),
+            buttonClass
+          )
+
+          expect(buttonHiddenAttribute).toBeFalsy()
+        })
+      })
+
+      describe('when the toggle button is clicked once', () => {
+        beforeAll(async () => {
+          await render(page, 'password-input', examples.default)
+          await page.click(buttonClass)
+        })
+
+        it('changes the input to type="text"', async () => {
+          const inputType = await page.evaluate(
+            (inputClass) =>
+              document.querySelector(inputClass).getAttribute('type'),
+            inputClass
+          )
+
+          expect(inputType).toEqual('text')
+        })
+
+        it('changes the status to say the password is visible', async () => {
+          const statusText = await page.evaluate(
+            (statusClass) =>
+              document.querySelector(statusClass).innerHTML.trim(),
+            statusClass
+          )
+
+          expect(statusText).toEqual('Your password is visible')
+        })
+
+        it('changes the button text to "hide"', async () => {
+          const buttonText = await page.evaluate(
+            (buttonClass) =>
+              document.querySelector(buttonClass).innerHTML.trim(),
+            buttonClass
+          )
+
+          expect(buttonText).toEqual('Hide')
+        })
+
+        it('changes the button aria-label to "hide password"', async () => {
+          const buttonAriaLabel = await page.evaluate(
+            (buttonClass) =>
+              document.querySelector(buttonClass).getAttribute('aria-label'),
+            buttonClass
+          )
+
+          expect(buttonAriaLabel).toEqual('Hide password')
+        })
+      })
+
+      describe('when the toggle button is clicked twice', () => {
```

Comment:
**question** Would it be worth testing that the password component can be made visible again after having been reverted to 'password', thinking of how some bugs sometimes happen when a component seemingly reverts back to its original state after interactions, but listeners or whatnot are not quite set up properly and future interaction breaks.

Reply:
If useful, this could be sorted by:

- gathering the `it` calls for the 'visible' and 'hidden' state of the password into two functions: `itShowsThePassword` and `itHidesThePassword`
- using [`describe.each`](https://jestjs.io/docs/api#describeeachtablename-fn-timeout) to run the tests against a given number of clicks and check that the component is in the right state

Something along the lines of (untested, so may not work).

```js
function itShowsThePassword() {
  // Tests for each part of the state (input type, button label,...)
	it(...) {
	}
	
	it(...) {
  }
}

function itHidesThePassword() {
	// Same but for when the password is hidden
}

describe.each([
	[1, itShowsThePassword],
  [2, itHidesThePassword],
  [3, itShowsThePassword]
])('when clicked %i time(s)', (clicks, expectation)  => {
	beforeAll(async () => {
    await render(page, 'password-input', examples.default)
    for (i = 0; i< clicks; i++) {
      await page.click(buttonClass)
    }
  })

  expectation()
}) 
```